File: cpp-linux/personalize/personalize.cpp
// personalize -- C++ port of examples\Vb6\personalize\personalize.bas
// Imports a tax form, fills in the fields and adds a web link.
#include <lumaspdf.h>
#include "repo_root.h"
#include <cstdio>
#include <string>
#include <ctime>

#ifndef RGB
#define RGB(r,g,b) ((UI32)(((unsigned char)(r))|(((unsigned char)(g))<<8)|(((unsigned char)(b))<<16)))
#endif

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return -1; // break processing on error
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfCreateNewPDFA(pdf, "");

    pdfSetViewerPreferences(pdf, vpDisplayDocTitle, avNone);
    pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage);
    const char* inFile = LUMAS_REPO_ROOT "/sample_form.pdf";
    if(pdfOpenImportFileA(pdf, inFile, ptOpen, "") < 0){ pdfDeletePDF(pdf); return 1; }
    pdfImportPDFFile(pdf, 1, 1.0, 1.0);

    pdfEditPage(pdf, 1);
    pdfSetFontA(pdf, "Courier", fsBold, 14.0, 0, cp1252);
    pdfWriteTextA(pdf, 72.5, 748.5, "X");
    pdfWriteTextA(pdf, 74.0, 701.0, "Musterstadt");
    pdfWriteTextA(pdf, 74.0, 677.0, "252/1062/3323");
    pdfBeginContinueText(pdf, 74.0, 628.0);
    pdfSetLeading(pdf, 24.0);
    pdfSetCharacterSpacing(pdf, 5.8);
    pdfAddContinueTextA(pdf, "Mustermann");
    pdfAddContinueTextA(pdf, "Hermann");
    pdfAddContinueTextA(pdf, "22021963keineKaufmann");
    pdfAddContinueTextA(pdf, "Musterstra\xDF" "e 145");   // 0xDF = 'sharp s' in cp1252 (split ends the hex escape)
    pdfAddContinueTextA(pdf, "12345Musterstadt");
    pdfSetCharacterSpacing(pdf, 0.0);
    pdfSetFontA(pdf, "Courier", fsBold, 10.0, 0, cp1252);
    pdfSetLeading(pdf, 48.0);
    pdfAddContinueTextA(pdf, "04.05.1994");
    pdfSetFontA(pdf, "Courier", fsBold, 14.0, 0, cp1252);
    pdfSetCharacterSpacing(pdf, 5.8);
    pdfAddContinueTextA(pdf, "Sabine");
    pdfSetLeading(pdf, 47.5);
    pdfAddContinueTextA(pdf, "18121966 ev  Hausfrau");
    pdfEndContinueText(pdf);
    pdfWriteTextA(pdf, 72.5, 365.0, "X");
    pdfWriteTextA(pdf, 396.0, 365.0, "X");
    pdfBeginContinueText(pdf, 74.0, 316.0);
    pdfSetLeading(pdf, 24.0);
    pdfAddContinueTextA(pdf, "2346256780     76834560");
    pdfAddContinueTextA(pdf, "Sparkasse Musterstadt");
    pdfEndContinueText(pdf);
    pdfWriteTextA(pdf, 72.5, 269.0, "X");
    pdfSetCharacterSpacing(pdf, 0.0);
    pdfSetFontA(pdf, "Courier", fsNone, 10.0, 0, cp1252);

    char nowStr[64];
    time_t t = time(nullptr);
    struct tm* lt = localtime(&t);
    strftime(nowStr, sizeof(nowStr), "%d.%m.%Y %H:%M:%S", lt);
    pdfWriteTextA(pdf, 53.0, 48.0, nowStr);

    pdfSetFillColor(pdf, RGB(0xFF, 0x66, 0x66));
    pdfSetFontA(pdf, "Helvetica", fsBold, 22.0, 0, cp1252);
    pdfWriteTextA(pdf, 340.0, 70.0, "www.lumaspdf.com");
    pdfSetLineWidth(pdf, 0.0);
    pdfSetLinkHighlightMode(pdf, hmPush);
    pdfSetAnnotFlags(pdf, afReadOnly);
    pdfWebLinkA(pdf, 340.0, 64.0, 204.0, 22.0, "https://www.lumaspdf.com");
    pdfEndPage(pdf);

    std::string outFile = exeDir(argv[0]) + "/out.pdf";
    if(pdfHaveOpenDoc(pdf) != 0){
        pdfSetOnErrorProc(pdf, nullptr, nullptr); // silence errors while opening the output file
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){ pdfDeletePDF(pdf); return 1; }
        pdfSetOnErrorProc(pdf, nullptr, ErrProc);
    }
    if(pdfCloseFile(pdf) != 0)
        printf("PDF file \"%s\" successfully created!\n", outFile.c_str());
    pdfDeletePDF(pdf);
    return 0;
}

File: vbnet/personalize/personalize.vb
' personalize -- VB.NET port of examples\Vb6\personalize
Imports System
Imports LumasPdfSdk

Module modPersonalize
    Private errDel As TErrorProc

    Private Function U(ByVal v As Long) As UInteger
        Return CUInt(v And &HFFFFFFFFL)
    End Function

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    Public Function ErrProc(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Console.WriteLine(ErrMessage)
        Return -1
    End Function

    Sub Main()
        errDel = New TErrorProc(AddressOf ErrProc)
        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        LumasPdf.pdfCreateNewPDFW(pdf, "")

        LumasPdf.pdfSetViewerPreferences(pdf, LumasPdfConsts.vpDisplayDocTitle, LumasPdfConsts.avNone)
        LumasPdf.pdfSetImportFlags(pdf, U(LumasPdfConsts.ifImportAll Or LumasPdfConsts.ifImportAsPage))
        If LumasPdf.pdfOpenImportFileW(pdf, AppPath() & "\taxform.pdf", LumasPdfConsts.ptOpen, "") < 0 Then
            LumasPdf.pdfDeletePDF(pdf)
            Return
        End If
        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0)

        LumasPdf.pdfEditPage(pdf, 1)
        LumasPdf.pdfSetFontW(pdf, "Courier", LumasPdfConsts.fsBold, 14.0, False, TCodepage.cp1252)
        LumasPdf.pdfWriteTextW(pdf, 72.5, 748.5, "X")
        LumasPdf.pdfWriteTextW(pdf, 74.0, 701.0, "Musterstadt")
        LumasPdf.pdfWriteTextW(pdf, 74.0, 677.0, "252/1062/3323")
        LumasPdf.pdfBeginContinueText(pdf, 74.0, 628.0)
        LumasPdf.pdfSetLeading(pdf, 24.0)
        LumasPdf.pdfSetCharacterSpacing(pdf, 5.8)
        LumasPdf.pdfAddContinueTextW(pdf, "Mustermann")
        LumasPdf.pdfAddContinueTextW(pdf, "Hermann")
        LumasPdf.pdfAddContinueTextW(pdf, "22021963keineKaufmann")
        ' ChrW, not Chr. Chr resolves its argument through the CURRENT
        ' ANSI codepage, and .NET 8 ships no CP1252 provider by default, so
        ' it throws "No data is available for encoding 1252" the moment this
        ' example is built against the NuGet package instead of .NET
        ' Framework 4.8. ChrW takes a Unicode code point directly. For every
        ' Chr(n) in this tree the two agree exactly: none used 128..159, the
        ' only range where CP1252 and Unicode differ (223 = U+00DF, "ss").
        LumasPdf.pdfAddContinueTextW(pdf, "Musterstra" & ChrW(223) & "e 145")
        LumasPdf.pdfAddContinueTextW(pdf, "12345Musterstadt")
        LumasPdf.pdfSetCharacterSpacing(pdf, 0.0)
        LumasPdf.pdfSetFontW(pdf, "Courier", LumasPdfConsts.fsBold, 10.0, False, TCodepage.cp1252)
        LumasPdf.pdfSetLeading(pdf, 48.0)
        LumasPdf.pdfAddContinueTextW(pdf, "04.05.1994")
        LumasPdf.pdfSetFontW(pdf, "Courier", LumasPdfConsts.fsBold, 14.0, False, TCodepage.cp1252)
        LumasPdf.pdfSetCharacterSpacing(pdf, 5.8)
        LumasPdf.pdfAddContinueTextW(pdf, "Sabine")
        LumasPdf.pdfSetLeading(pdf, 47.5)
        LumasPdf.pdfAddContinueTextW(pdf, "18121966 ev  Hausfrau")
        LumasPdf.pdfEndContinueText(pdf)
        LumasPdf.pdfWriteTextW(pdf, 72.5, 365.0, "X")
        LumasPdf.pdfWriteTextW(pdf, 396.0, 365.0, "X")
        LumasPdf.pdfBeginContinueText(pdf, 74.0, 316.0)
        LumasPdf.pdfSetLeading(pdf, 24.0)
        LumasPdf.pdfAddContinueTextW(pdf, "2346256780     76834560")
        LumasPdf.pdfAddContinueTextW(pdf, "Sparkasse Musterstadt")
        LumasPdf.pdfEndContinueText(pdf)
        LumasPdf.pdfWriteTextW(pdf, 72.5, 269.0, "X")
        LumasPdf.pdfSetCharacterSpacing(pdf, 0.0)
        LumasPdf.pdfSetFontW(pdf, "Courier", LumasPdfConsts.fsNone, 10.0, False, TCodepage.cp1252)
        LumasPdf.pdfWriteTextW(pdf, 53.0, 48.0, Now().ToString())
        LumasPdf.pdfSetFillColor(pdf, U(&HFF Or (&H66 << 8) Or (&H66 << 16)))
        LumasPdf.pdfSetFontW(pdf, "Helvetica", LumasPdfConsts.fsBold, 22.0, False, TCodepage.cp1252)
        LumasPdf.pdfWriteTextW(pdf, 340.0, 70.0, "www.lumaspdf.com")
        LumasPdf.pdfSetLineWidth(pdf, 0.0)
        LumasPdf.pdfSetLinkHighlightMode(pdf, CInt(THighlightMode.hmPush))
        LumasPdf.pdfSetAnnotFlags(pdf, U(LumasPdfConsts.afReadOnly))
        LumasPdf.pdfWebLinkW(pdf, 340.0, 64.0, 204.0, 22.0, "https://www.lumaspdf.com")
        LumasPdf.pdfEndPage(pdf)

        Dim outFile As String = AppPath() & "\out.pdf"
        If LumasPdf.pdfHaveOpenDoc(pdf) <> 0 Then
            LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, Nothing)
            If LumasPdf.pdfOpenOutputFileW(pdf, outFile) = 0 Then
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
            LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, errDel)
        End If
        If LumasPdf.pdfCloseFile(pdf) <> 0 Then
            Console.WriteLine("PDF file """ & outFile & """ successfully created!")
        End If
        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module
